### PR TITLE
Tar: support GNU numeric format.

### LINKED ISSUE #93763: TarReader throws on archive that other tools accept
### Description

I have a *.tar.gz file that I am trying to process in C# by passing the GZipStream to TarReader.

This fails at the line 

TarHelpers.ParseOctal<ulong>(buffer.Slice(124, 12))

The 

if (num >= 8)

check inside ParseOctal is getting to the ThrowInvalidNumber part as the num being tested is 80.

I have seen a similar issue where non conformance to the spec is blamed for this type of thing.

But Azure Data Factory can process this *.tar.gz fine and additionally if I extract the *.tar then Windows Explorer seems able to deal with it and shows it as a compressed archive - clicking through shows the details of the single file contained within. 

The complete 512 bytes of the header is 

```
6E625F726D73655F727061735F696E76656E746F72792E64617400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003030303036363600303135323036310030313532303631008000000000000011E7310C9D3134353134303736373333003031363332350020300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000075737461722020006F7261636C6500000000000000000000000000000000000000000000000000006F696E7374616C6C000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

So the problematic part is 

`8000000000000011E7310C9D`

The underlying file size is 76893195421 bytes.

### Reproduction Steps

Not practical for me to share the actual file but it consists of the 512 byte header already given. 

Followed by 76893195421 bytes of arbitrary data that is the actual file contents

Followed by a new line and 8547 null bytes.

You can use 
```
var header = Convert.FromHexString(
    "6E625F726D73655F727061735F696E76656E746F72792E64617400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003030303036363600303135323036310030313532303631008000000000000011E7310C9D3134353134303736373333003031363332350020300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000075737461722020006F7261636C6500000000000000000000000000000000000000000000000000006F696E7374616C6C000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000");


System.IO.File.WriteAllBytes(@"C:\somefile.tar", header
    );
```

And then open the archive in 7zip and see that this utility can cope with it as one example

image

### Expected behavior

These types of file have been generated by a system in my company for years.

Every utility used to read them has apparently had no issues with them except for the new .NET framework classes.

I have no idea whether or not this file format does in fact violate any spec but I would expect to be able to open it as other utilities can cope (maybe by setting a mode parameter to ignore such issues as long as they aren't fatal to the extraction)

### Actual behavior

System.IO.InvalidDataException: 'Unable to parse number.' exception thrown on initial tarReader.GetNextEntry() call

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by tmds**
The tar specification stores numeric fields using an octal representation. This limits the range of values that can be stored.

To increase the supported range, a GNU extension defines that when the leading byte is 0xff/0x80 the remaining bytes are a negative/positive big endian formatted value.

When writing under the PAX format, we continue to only use the only octal representation in the header fields. The values are overridden using extended attributes.

Fixes https://github.com/dotnet/runtime/issues/93763.

@dotnet/area-system-formats-tar ptal.

### CODE REVIEW

**Comment by tmds** [comment]
Test failures are unrelated.

**Comment by tmds** [comment]
@carlossanlop ptal.

**Comment by am11** [comment]
It would be nice to get this in for .NET 9. The sooner tar implementation matures the better for the ecosystem. Thanks for this @tmds! 👍

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:101]: For `TarEntry.ModificationTime` and `PosixTarEntry.DeviceMinor` and `PosixTarEntry.DeviceMajor`:

I think we should still throw if the current entry's tar entry format is _not_ GNU and the value is beyond this limit for octal.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/GnuTarEntry.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/GnuTarEntry.cs:115]: Should we have a different lower limit here?

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:20]: Maybe we could use these limits in the PosixTarEntry and TarEntry setter exceptions for ModificationTime, DeviceMajor and DeviceMinor, when the tar entry format is not GNU. Thoughts?

The higher value maybe does not need to throw for PAX anymore, as you are storing the value in the extended attributes when it does not fit in the standard field.

File: src/libraries/System.Formats.Tar/tests/Manual/ManualTests.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/tests/Manual/ManualTests.cs:26]: I'm not sure I don't understand why this was moved to the forloop above, can you please explain?

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs:266]: Why was this removed?

**Review by carlossanlop — COMMENTED** [review]
Thanks for this change. I left some questions. Haven't gone through the whole PR yet.

We also need to run runtime-extra-platforms tests.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (tmds) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:101]: > I think we should still throw if the current entry's tar entry format is not GNU and the value is beyond this limit for octal.

That makes sense. I checked and the GNU `tar` tool also doesn't want to use the GNU numeric format when using a non-GNU tar format. For reading it does accept the GNU numeric format independent of the archive tar format. I don't think we need to be as liberal. We don't expect to see the format in non-GNU archives, and it would require special handling in the tests because we wouldn't be able to use the writer to create the archives.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/GnuTarEntry.cs
Comment (tmds) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/GnuTarEntry.cs:115]: I'll update this per https://github.com/dotnet/runtime/pull/101172#discussion_r1609681536.

With the GNU numeric format there will be no lower range. With the other formats it will throw for less than `UnixEpoch` as before.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (tmds) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:20]: > The higher value maybe does not need to throw for PAX anymore, as you are storing the value in the extended attributes when it does not fit in the standard field.

Yes, we can also allow the larger range for the PAX format. I'll include this.

File: src/libraries/System.Formats.Tar/tests/Manual/ManualTests.cs
Comment (tmds) [reply src/libraries/System.Formats.Tar/tests/Manual/ManualTests.cs:26]: Before the change, only the Pax format allowed the larger files. This test was updated to run against all formats.

Based on the other feedback, we'll allow the larger file range for Pax and Gnu, and I'll update this accordingly.

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs
Comment (tmds) [reply src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs:266]: The size field of `writer-big` is a valid GNU numeric format. The issue with the file is that it doesn't include all data for that large file size, which throws as the reader moves across the data, but not when reading the header (which is what the test is doing).

**Comment by tmds** [comment]
@carlossanlop I have addressed your feedback, can you take another look?

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:1092]: I'm not sure I get why we also have this 32BigEndian call. Can you please describe it for my own education?

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/GnuTarEntry.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/GnuTarEntry.cs:115]: Ok I see what you did here. Since aTime and cTime properties are only available in the GNU format, we don't need to throw for this specific UnixEpoch case, as we should be able to set these fields to MinValue.

And for PAX, those two fields get added directly to the extended attributes, there are no properties for them.

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs:266]: This is one of those files that we found in a golang tar reader and decided to bring it here to ensure we had a clearly documented behavior in our reader. This case had an unexpected char in one of the fields so we _expect_ our tar reader to throw.

So based on your description, GNU can advance a little more than the other formats, but the file is _still_ malformed even for GNU. If you re-add the test case, does it still throw InvalidDataException? If yes, then I think we should bring that test case back. If not, then I think we should duplicate this test method just for that test case and make sure we capture the exception that is thrown.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/PosixTarEntry.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/PosixTarEntry.cs:65]: Just to be clear that I get this: Are DeviceMajor and DeviceMinor values greater than 0x1FFFFF valid in Unix?

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/PosixTarEntry.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/PosixTarEntry.cs:62]: Can't comment above this line: Can you please also update the exception message in line 53?

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.cs:101]: Can't comment above this line: Can you please also update the exception message in line 95?

File: src/libraries/System.Formats.Tar/tests/Manual/ManualTests.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/tests/Manual/ManualTests.cs:30]: Can't write below this line: Do you mind running the manual test in line 34? `WriteEntry_LongFileSize`. It's one of those tests that are difficult to run in CI.

File: src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.Base.cs
Comment (carlossanlop) [inline src/libraries/System.Formats.Tar/tests/TarWriter/TarWriter.WriteEntry.Base.cs:87]: Low pri nit, and I know this is slightly beyond your PR, but if you get a chance, can you please rename this method (and all the invocations) to reflect that this is just for PAX? `WriteTimeStamp_Pax_TheoryData` or something like that.

**Review by carlossanlop — COMMENTED** [review]
Thank you for addressing the feedback. I gave it another pass and left some comments for you to consider.

I'll run runtime-extra-platforms now to see if we catch any unexpected errors.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/PosixTarEntry.cs
Comment (tmds) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/PosixTarEntry.cs:65]: Yes. Unix/POSIX says that you can turn a `dev_t` into a major and a minor and vice versa. It doesn't put a limit on the size of `dev_t` or say how many bits are used by the major and how many by the minor.

You can look at a concrete implementation (like a Linux kernel) and determine what its limits are, but there is no standard that sets an upper bound.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (tmds) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs:1092]: We're filling a 12 byte field. The lower 8 bytes are the long. We need to use the appropriate value for the first 4 bytes depending on whether the value is negative or positive. I'm going to replace `1U << 31` with `0x80000000` as that better matches the comment above.

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs
Comment (tmds) [reply src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs:266]: This file is a header that describes a 16GB file. Previously it failed because the size was considered invalid.
Now it fails because we're trying to seek past the length.

> If not, then I think we should duplicate this test method just for that test case and make sure we capture the exception that is thrown.

The test would be something like this:

```cs
        [Fact]
        public void Throw_ArchiveIsShort()
        {
            // writer-big has a header for a 16G file but not its contents.
            using MemoryStream archiveStream = GetTarMemoryStream(CompressionMethod.Uncompressed, "golang_tar", "writer-big");
            using TarReader reader = new TarReader(archiveStream);
            // MemoryStream throws when we try to change its Position past its length.
            Assert.Throws<ArgumentOutOfRangeException>(() => reader.GetNextEntry());
        }
```

Should I add it?

File: src/libraries/System.Formats.Tar/tests/Manual/ManualTests.cs
Comment (tmds) [reply src/libraries/System.Formats.Tar/tests/Manual/ManualTests.cs:30]: I ran the manual test in this project, and they all passed.

File: src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs
Comment (tmds) [reply src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Write.cs]: I'm not sure what comment you mean. The `int` values can always be stored in the 8-byte field using the GNU format. It supports the entire range from `int.MinValue` to `int.MaxValue`.

This code extends the int to a long. Because it is 2-complement, negative values are extended with all 0xff.
For positive we need to set the highest bit. We don't do it conditionally on `value` (to avoid a branch) because also for negative numbers the highest bit is set.

**Comment by tmds** [comment]
@carlossanlop I've addressed the additional feedback except for the `writerbig` test. See https://github.com/dotnet/runtime/pull/101172#discussion_r1629500345.

**Comment by tmds** [comment]
@carlossanlop can you let me know if you want this test included: https://github.com/dotnet/runtime/pull/101172#discussion_r1629500345?

File: src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs
Comment (carlossanlop) [reply src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs:266]: Yes please! Let's not lose coverage of it.

**Review by carlossanlop — APPROVED** [review]
Approved pending two things:

- The re-addition of the removed writer-big test case as suggested in the snippet.
- I ran runtime-extra-platforms, which executes tar tests in many other platforms (mobile maybe). Please verify we don't see any _new_ failures related to this change.

**Comment by tmds** [comment]
@carlossanlop can you start another run of runtime-extra-platforms?

**Review by carlossanlop — APPROVED** [review]

**Comment by tmds** [comment]
@carlossanlop this is good to merge.

**Comment by carlossanlop** [comment]
Thanks @tmds!

**Comment by carlossanlop** [comment]
Considering this is a problem that [can be seen in runfo](https://github.com/dotnet/runtime/issues/103734) (one of our infrastructure tools) I'm inclined to backport this to .NET 8. Would you have any objections, @ericstj?

**Comment by ericstj** [comment]
Runfo isn't a critical tool, its just for debugging.  I'd instead suggest you seek out input from customers reporting this one.  See if anyone is blocked on .NET 8.0 - that's more important.   This definitely would meet the bar for servicing if we had a customer blocked by it.

**Pull request merged**
